Check output of simulations

orelgueta opened the pull request:
The goal of this PR is to introduce an additional step to the integration tests of the simulation productions that checks the output and makes sure it is reasonable.

The goal above required various changes throughout the code: 
- allow setting random seeds in order to produce reproducible results
- add parameters and schemas that are necessary to reproduce the Prod6 results
- change hardcoded values that did not comply with the Prod6 model

There are also a few random bits of changes introduced as a part of this PR which are not directly related to the subject, but I tried to make sure the commit messages clarify those in case they need clarification.

This PR is dependent on the [corresponding MR](https://gitlab.cta-observatory.org/cta-science/simulations/simulation-model/model_parameters/-/merge_requests/60) in the model parameters repo. Tests will fail until that one is merged.

Fixes #1129.

GernotMaier commented on simtools/corsika/corsika_config.py:
```diff
@@ -52,7 +52,8 @@ def __init__(self, array_model, args_dict, db_config=None, label=None):
         self.config_file_path = None
-        self.primary_particle = self._set_primary_particle(args_dict)
+        # The following uses the setter defined below, that is why the args_dict is passed
+        self.primary_particle = args_dict
```
This is correct but is without reading the functions below very confusing. Not nice that an additional comment is needed.

orelgueta replied:
What do you suggest we do then? The other option I see to make it more readable is to change to
```
self.primary_particle = {"primary_id_type": args_dict.get("primary_id_type"), "primary": args_dict.get("primary")}
```

But I think this is worse than the comment explaining it to future developers.

GernotMaier replied:
No suggestion - I was just complaining. Should have made it clearer that no changes are expected.

GernotMaier commented on simtools/model/model_parameter.py:
Do you actually remember why we had this 'or' statement? There must have been a reason.

orelgueta replied:
Yes, it was there in case no label was provided to the application. That caused an issue actually because only in the case of the model parameter the output went to a directory with the name of the telescope instead of a directory under the date (`d-19-09-24`). Then the application would fail because the model parameters were not found. 

This fixes that (though I know you don't like the date thing.

GernotMaier replied:
No. I don’t like the date thing - `use_plain_output_path` is my friend. Thanks for the explanation!

GernotMaier commented on simtools/model/array_model.py:
```diff
@@ -108,6 +108,11 @@ def _initialize(self, site, array_elements_config):
+        if array_elements == {}:
+            raise ValueError(
+                "No array elements found. "
+                "Did you provide a valid layout name or list of telescopes?"
```
I don't like the 'you' in the message - it implies that the user is a person. Maybe "Possibly missing valid layout name or missing telescope list"?

orelgueta replied:
No problem.

GernotMaier commented on simtools/model/site_model.py:
```diff
@@ -68,20 +68,22 @@ def get_corsika_site_parameters(self, config_file_style=False):
+                    self.get_parameter_value_with_unit("corsika_observation_level").to_value("cm")
+                ],
+                # We always use a custom profile by filename, so this has to be set to 99
+                "ATMOSPHERE": [99, "Y"],
```
Good. This closes #829

GernotMaier commented on simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml:
```diff
@@ -0,0 +1,38 @@
+  but non-zero values of this parameter will also activate save_photons bit 1.
+  Storing p.e. in internal calibration events is controlled separately by save_calibration_pe.
+data:
+  - type: int
```
Would `type: boolean` be more appropriate? We have done that with all (most?) other parameters with this type of range (I think Konrad uses this because bool was not available in his C version when he started sim_telarray).

orelgueta replied:
Changed to boolean

GernotMaier commented on simtools/schemas/model_parameters/save_pe_with_amplitude.schema.yml:
```diff
@@ -0,0 +1,38 @@
+  Storing p.e. in internal calibration events is controlled separately by save_calibration_pe.
+data:
+  - type: int
+    unit: byte
```
It this unit correct? I understood it as boolean.

orelgueta replied:
changed to boolean

GernotMaier commented on simtools/schemas/model_parameters/store_photoelectrons.schema.yml:
```diff
@@ -0,0 +1,41 @@
+  and produce any photo-electrons if there are too few photons in the first place.
+data:
+  - type: int
+    unit: byte
```
Is this unit correct?

orelgueta replied:
No, it's in p.e., but that's not a valid unit, so I set it to dimensionless.

GernotMaier commented on simtools/simtel/simulator_array.py:
```diff
@@ -111,3 +114,30 @@ def _check_run_result(self, run_number=None):
         self._logger.debug(f"simtel_array output file {output_file} exists.")
         return True
+
+    @staticmethod
```
Can you explain the choice of having those values hardwired? Any easy way to make them configurable?

orelgueta replied:
I hardcoded them in the same way they are hardcoded in sim_telarray. Since we do not use the sim_telarray histograms, I didn't see a reason to make it configurable. However, if you think it's worth it, please make this comment into an issue and I will implement it after this PR (and use this function to set default values).

GernotMaier replied:
I think it is easy to miss that this is for histogramming only. Could we either change the function call or at least say in the first line of the docstring `..used for histogramming in sim_telarray`? Quick readers might otherwise thing we hardwire the spectral index for the simulations.

orelgueta replied:
Changed the name of the function to get_power_law_for_sim_telarray_histograms.

GernotMaier commented on simtools/testing/assertions.py:
```diff
@@ -41,3 +43,109 @@ def assert_file_type(file_type, file_name):
+
+def assert_n_showers_and_energy_range(file):
+    """
+    Assert #showers and energy range.
```
What does `#` mean?

GernotMaier replied:
And maybe say that you look into a log file of corsika or simtelarray here?

orelgueta replied:
`#` means number and I think I wrote it because I didn't want to repeat myself with the word "number". I changed it back to number.
This is one of those cases in which I had to reformulate because ruff wouldn't let me have one long sentence which took more than one line...


I do not look into the log files of CORSIKA and sim_telarray here but instead read the sim_telarray output file itself and extract the headers and count the `#` of events and energies produced.

GernotMaier replied:
Still don’t understand why you can’t use `number`. The first line is not that long.

orelgueta replied:
I changed it already to be `Assert the number of showers and the energy range.`

My complaint was about the original version you didn't see which was 
```
Assert the number of showers is consistent with the required one (up to 1% tolerance) 
and that the energies simulated are within the configured ones.
```

This is too long to be one line and ruff doesn't allow a summary line with a line break in it, so I had to break it into two separate statements with some repetition. I find it a bit annoying, so I complained.

GernotMaier reviewed (approved):
Thanks a lot for this big update - really hard to keep up with all those little details, I wouldn't have known them.

My comments are mostly minor, this is why I approve already.

GernotMaier commented:
@orelgueta - I guess you have seen the sonarqu complaint about duplications:

https://sonar-cta-dpps.zeuthen.desy.de/component_measures?id=gammasim_simtools_AY_ssha9WiFxsX-2oy_w&metric=new_duplicated_lines_density&pullRequest=1139&view=list

orelgueta commented:
Thanks for the detailed review! 

I fixed the comments and answered a couple of points. I won't merge yet because of the point regarding the primary particle.

I also fixed the duplication, so I will wait for sonarcube to run again. I didn't know it is so pedantic, but it's not bad I guess.

orelgueta commented:
The integration tests are failing because of the change to the pixel rate that is missing in the db. The old style of the long string doesn't work anymore, it is replaced with a real list of values instead (which is shortened in the cfg file). We could wait for the model parameter MR to create a new DB, rerun the tests and only then merge this one. That would be the safest.

GernotMaier commented:
And yes, sonarq is pedantic and our ruff settings are strict. This is good as it really simplifies reviews and we are not bothered with this stuff.

GernotMaier commented on simtools/schemas/model_parameters/sampled_output.schema.yml:
```diff
@@ -0,0 +1,35 @@
+description: |-
+  Select a specific output format (if several available). A value of 0 means to write ADC sums, a value of 1 to write ADC samples.
+data:
+  - type: int
```
Should this be boolean?

orelgueta replied:
Yes, changed to boolean, I missed it yesterday.

GernotMaier commented on simtools/schemas/model_parameters/sampled_output.schema.yml:
```diff
@@ -0,0 +1,35 @@
+  Select a specific output format (if several available). A value of 0 means to write ADC sums, a value of 1 to write ADC samples.
+data:
+  - type: int
+    unit: byte
```
Should this be null?

orelgueta replied:
Removed since it is a boolean

orelgueta commented:
> Two minor comments, but as before already approved.
> 
> This is what I usually do for testing (and you might have done it):
> 
> * upload the branch in question of the model parameter repository to a local database
> * Run the integration tests locally (remember to run them with `—model_version=“6.0.0”` and `5.0.0`
> 
> If this is fine, we probably can first merge the model_parameter changes; release a new version there (to trigger the DB upload) and then re-run the integration tests here on github.

Yes I have done all of that locally during the development process. Otherwise there's no way to test anything. It should all work now after the merging of both branches.

orelgueta commented:
I resolved the conflicts with the main branch. I will wait for the tests to pass (or perhaps fail if the DB is not updated yet) and then merge.
